onecodex/helpers.py

```diff
@@ -1,169 +1,328 @@
-    """
-    normed_classifications = []
-    metadata = []
-
-    DEFAULT_FIELDS = list(Metadata._resource._schema['properties'].keys())
-    DEFAULT_FIELDS.remove('$uri')
-    DEFAULT_FIELDS.remove('sample')
-
-    for a in analyses:
-        if isinstance(a, Samples):
-            c = a.primary_classification
-            m = a.metadata
-        elif isinstance(a, Classifications):
-            c = a
-            m = a.sample.metadata
-        elif isinstance(a, Analyses):
-            if a.analysis_type != 'classification':
-                raise OneCodexException('{} is not a classification'.format(a.id))
-            c = Classifications(a._resource._client.Classifications.fetch(a.id))
-            m = a.sample.metadata
-
-        if skip_missing and not c.success:
-            if warn:
-                warnings.warn('Classification {} not successful. Skipping.'.format(c.id))
-            continue
-
-        normed_classifications.append(c)
-        metadatum = {f: getattr(m, f) for f in DEFAULT_FIELDS}
-
-        # Add sample_id, metadata_id, and sample created_at
-        metadatum['sample_id'] = m.sample.id
-        metadatum['metadata_id'] = m.id
-        metadatum['created_at'] = m.sample.created_at
-
-        if label is None:
-            metadatum['_display_name'] = (metadatum['name'] if metadatum['name'] is not None
-                                          else c.sample.filename)
-        elif isinstance(label, six.string_types):
-            if label in metadatum:
-                metadatum['_display_name'] = metadatum[label]
-            elif label in m.custom:
-                metadatum['_display_name'] = m.custom[label]
-            else:
-                metadatum['_display_name'] = None
-        elif callable(label):
-            metadatum['_display_name'] = label(m)
-        else:
-            raise NotImplementedError('Must pass a string or lambda function to `label`.')
-
-        metadatum.update(m.custom)
-        metadata.append(metadatum)
-
-    metadata = pd.DataFrame(metadata)
-    if all(pd.isnull(metadata['_display_name'])):
-        raise OneCodexException('Could not find any labels for `{}`'.format(label))
-
-    return normed_classifications, metadata
+from onecodex.viz import VizPCAMixin, VizHeatmapMixin, VizMetadataMixin, VizDistanceMixin
+
+
+# force persistence of our additional taxonomy and metadata dataframe properties
+class ResultsDataFrame(pd.DataFrame):
+    _metadata = ['ocx_rank', 'ocx_field', 'ocx_taxonomy', 'ocx_metadata']
+
+    def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False, ocx_data={}):
+        self.ocx_rank = ocx_data.get('ocx_rank', None)
+        self.ocx_field = ocx_data.get('ocx_field', None)
+        self.ocx_taxonomy = ocx_data.get('ocx_taxonomy', None)
+        self.ocx_metadata = ocx_data.get('ocx_metadata', None)
+
+        pd.DataFrame.__init__(self, data=data, index=index, columns=columns, dtype=dtype, copy=copy)
```

Comment:
I'm not sure what this line is doing. `__init__` has to return `None` and this DataFrame isn't being assigned to anything.

Reply:
It's equivalent to `super(ResultsDataFrame, self).__init__()` in Python2-style syntax, or `super().__init__()` in Python3 style. The reason I used this instead of `super()` is that ultimately `ResultsDataFrame` will inherit from multiple classes and I wanted to be explicit about which parent class's `__init__` I wanted to call.